File: rp/dictionary/three_sequence_dictionary.py
"""used for typing method arguments"""
from typing import List, Dict
import re

class ThreeSequenceDictionary:
    """Dictionary that saves three word sequences instead of only 2"""
    def __init__(self):
        super().__init__()
        self.dictionary: List[str] = []
        self.coordinates: Dict[str, List[int]] = {}

    def load(self, source_file: str, coordinates_file: str) -> None:
        """loads the dictionary from source files"""
        with open(source_file, "r", encoding="ascii") as f1:
            for line in f1:
                for word in line.split():
                    word = re.sub(r"[^A-Za-z\-]", '', word)
                    self.dictionary.append(word.lower())

        with open(coordinates_file, "r", encoding="ascii") as f2:
            for line in f2:
                words = line.split()
                self.coordinates[words[0]] = list(map(int, words[1:]))

    def contains(self, middle_word: str, first_last: List[str]) -> bool:
        """returns if there is an entry with first_word where next_words are the following words in sequence"""
        middle_word = re.sub(r"[^A-Za-z\-]", '', middle_word).lower()
        first_last = list(map(lambda x: re.sub(r"[^A-Za-z\-]", '', x).lower(), first_last))
        if middle_word not in self.coordinates: return False

        coordinates = self.coordinates.get(middle_word, [])
        for coord in coordinates:
            if 0 < coord < len(self.dictionary) - 1:
                if (self.dictionary[coord - 1] == first_last[0] and
                        self.dictionary[coord + 1] == first_last[1]):
                    return True
        return False

File: rp/text/dictionary/three_sequence_dictionary.cpp
#include "three_sequence_dictionary.hpp"
#include "../text_util.hpp"
#include <fstream>

void ThreeSequenceDictionary::load(const std::string& source_file, const std::string& coordinates_file) {
    {
        std::ifstream f1(source_file);
        std::string line;
        while (std::getline(f1, line)) {
            for (const std::string& word : split_words(line)) {
                dictionary.push_back(clean_word(word));
            }
        }
    }

    std::ifstream f2(coordinates_file);
    std::string line;
    while (std::getline(f2, line)) {
        std::vector<std::string> words = split_words(line);
        if (words.empty()) continue;
        std::vector<int> coords;
        coords.reserve(words.size() - 1);
        for (std::size_t i = 1; i < words.size(); ++i) coords.push_back(std::stoi(words[i]));
        coordinates[words[0]] = std::move(coords);
    }
}

bool ThreeSequenceDictionary::contains(const std::string& middle_word_raw,
                                       const std::vector<std::string>& first_last_raw) const {
    const std::string middle_word = clean_word(middle_word_raw);
    std::vector<std::string> first_last;
    first_last.reserve(first_last_raw.size());
    for (const std::string& w : first_last_raw) first_last.push_back(clean_word(w));

    auto it = coordinates.find(middle_word);
    if (it == coordinates.end()) return false;

    for (int coord : it->second) {
        if (coord > 0 && coord < static_cast<int>(dictionary.size()) - 1) {
            if (dictionary[coord - 1] == first_last[0] && dictionary[coord + 1] == first_last[1]) {
                return true;
            }
        }
    }
    return false;
}
